tiledb/sm/query/query.cc

```diff
@@ -560,832 +569,42 @@ tiledb::sm::Status Query::from_capnp(rest::capnp::Query::Reader* query) {
-                    std::to_string(buffer_length) + ")");
-              }
-              char* existing_buffer_casted = static_cast<char*>(existingBuffer);
-              for (uint i = 0; i < buffer_length; i++) {
-                existing_buffer_casted[i] = newBuffer[i];
-              }
-
-              if (buffer.hasBufferOffset()) {
-                auto buffer_offset = buffer.getBufferOffset();
-                if (*existingBufferOffsetSize / sizeof(uint64_t) !=
-                    buffer_offset.size()) {
-                  return Status::QueryError(
-                      "Existing buffer_var_ in query object is different size "
-                      "(" +
-                      std::to_string(*existingBufferOffsetSize) +
-                      ") vs new query object buffer_var size (" +
-                      std::to_string(buffer_offset.size()) + ")");
-                }
-                uint64_t* existing_buffer_var_caster =
-                    static_cast<uint64_t*>(existingBufferOffset);
-                for (uint i = 0; i < buffer_offset.size(); i++) {
-                  existing_buffer_var_caster[i] = buffer_offset[i];
-                }
-              }
-
-            } else {
-              // Get buffer
-              std::vector<char>* new_buffer =
-                  new std::vector<char>(buffer_length);
-              for (uint i = 0; i < buffer_length; i++) {
-                (*new_buffer)[i] = newBuffer[i];
-              }
-              uint64_t* buffer_size = new uint64_t(buffer_length * type_size);
-
-              // Check for offset buffer
-              if (buffer.hasBufferOffset()) {
-                auto buffer_offset = buffer.getBufferOffset();
-                std::vector<uint64_t>* new_buffer_offset =
-                    new std::vector<uint64_t>(buffer_offset.size());
-                for (uint i = 0; i < buffer_offset.size(); i++) {
-                  (*new_buffer_offset)[i] = buffer_offset[i];
-                }
-
-                uint64_t* new_buffer_offset_size =
-                    new uint64_t(new_buffer_offset->size() * sizeof(uint64_t));
-                set_buffer(
-                    attr->name(),
-                    new_buffer_offset->data(),
-                    new_buffer_offset_size,
-                    new_buffer->data(),
-                    buffer_size);
-              } else
-                set_buffer(attr->name(), new_buffer->data(), buffer_size);
-            }
-          }
-          break;
-        }
-        case tiledb::sm::Datatype::ANY: {
-          // Not supported datatype for buffer
-          return Status::Error(
-              "Any datatype not supported for deserialization");
-          break;
-        }
+      auto offset_buffer_size_in_bytes = buffer.getOffsetBufferSizeInBytes();
+      if (offset_buffer_size_in_bytes) {
+        // variable size attribute buffer
+        auto buffer_size_in_bytes = buffer.getBufferSizeInBytes();
+        set_buffer(
```

Comment:
As @Shelnutt2 pointed out we should be doing a memcpy here instead of `set_buffer`.

Reply:
@tdenniston on the server we will not have allocated the query buffers, I think in that case we do need a set_buffer call (to avoid having to have the server init buffers and the memcpy server side). So the ideal logic in my mind at this stage would be. Check if query object has buffers, if it does memcpy into them, if it doesn't call set_buffer. What do you think?

Reply:
I was wondering how to differentiate between memcpy and zero copy versions of buffer handling, i would have asked anyway. I we all agree that this is the way i can proceed.

Reply:
Ok, please proceed. Possibly this logic will be revisited anyway when I tackle incomplete queries.